File: app/src/main/java/io/nekohasekai/sagernet/database/SubscriptionBean.java
package io.nekohasekai.sagernet.database;

import androidx.annotation.NonNull;

import com.esotericsoftware.kryo.io.ByteBufferInput;
import com.esotericsoftware.kryo.io.ByteBufferOutput;

import io.nekohasekai.sagernet.fmt.Serializable;

public class SubscriptionBean extends Serializable {

    public static final Creator<SubscriptionBean> CREATOR = new CREATOR<SubscriptionBean>() {
        @NonNull
        @Override
        public SubscriptionBean newInstance() {
            return new SubscriptionBean();
        }

        @Override
        public SubscriptionBean[] newArray(int size) {
            return new SubscriptionBean[size];
        }
    };
    public Integer type;
    public String link;
    public String token;
    public Boolean forceResolve;
    public Boolean deduplication;
    public Boolean updateWhenConnectedOnly;
    public String customUserAgent;
    public Boolean autoUpdate;
    public Integer autoUpdateDelay;
    public Integer lastUpdated;

    // SIP008
    public Long bytesUsed;
    public Long bytesRemaining; // Also for OOC

    // Open Online Config
    public String username;
    public Long expiryDate;

    public SubscriptionBean() {
    }

    @Override
    public void serializeToBuffer(ByteBufferOutput output) {
        output.writeInt(2);

        output.writeInt(type);

        output.writeString(link);

        output.writeBoolean(forceResolve);
        output.writeBoolean(deduplication);
        output.writeBoolean(updateWhenConnectedOnly);
        output.writeString(customUserAgent);
        output.writeBoolean(autoUpdate);
        output.writeInt(autoUpdateDelay);
        output.writeInt(lastUpdated);
        output.writeLong(expiryDate);
        output.writeLong(bytesUsed);
        output.writeLong(bytesRemaining);
    }

    public void serializeForShare(ByteBufferOutput output) {
        output.writeInt(0);

        output.writeInt(type);

        output.writeString(link);

        output.writeBoolean(forceResolve);
        output.writeBoolean(deduplication);
        output.writeBoolean(updateWhenConnectedOnly);
        output.writeString(customUserAgent);
    }

    @Override
    public void deserializeFromBuffer(ByteBufferInput input) {
        int version = input.readInt();

        type = input.readInt();
        link = input.readString();
        forceResolve = input.readBoolean();
        deduplication = input.readBoolean();
        updateWhenConnectedOnly = input.readBoolean();
        customUserAgent = input.readString();
        autoUpdate = input.readBoolean();
        autoUpdateDelay = input.readInt();
        lastUpdated = input.readInt();
        if (version < 2) return;
        expiryDate = input.readLong();
        bytesUsed = input.readLong();
        bytesRemaining = input.readLong();
    }

    public void deserializeFromShare(ByteBufferInput input) {
        int version = input.readInt();

        type = input.readInt();
        link = input.readString();
        forceResolve = input.readBoolean();
        deduplication = input.readBoolean();
        updateWhenConnectedOnly = input.readBoolean();
        customUserAgent = input.readString();
    }

    @Override
    public void initializeDefaultValues() {
        if (type == null) type = 0;
        if (link == null) link = "";
        if (token == null) token = "";
        if (forceResolve == null) forceResolve = false;
        if (deduplication == null) deduplication = false;
        if (updateWhenConnectedOnly == null) updateWhenConnectedOnly = false;
        if (customUserAgent == null) customUserAgent = "";
        if (autoUpdate == null) autoUpdate = false;
        if (autoUpdateDelay == null) autoUpdateDelay = 1440;
        if (lastUpdated == null) lastUpdated = 0;

        if (bytesUsed == null) bytesUsed = 0L;
        if (bytesRemaining == null) bytesRemaining = 0L;

        if (username == null) username = "";
        if (expiryDate == null) expiryDate = 0L;
    }

}

File: app/src/main/kotlin/io/nekohasekai/sagernet/database/SubscriptionBean.kt
package io.nekohasekai.sagernet.database

import com.esotericsoftware.kryo.io.ByteBufferInput
import com.esotericsoftware.kryo.io.ByteBufferOutput
import io.nekohasekai.sagernet.fmt.Serializable

class SubscriptionBean : Serializable() {

    companion object {
        @JvmField
        val CREATOR = object : CREATOR<SubscriptionBean>() {
            override fun newInstance(): SubscriptionBean {
                return SubscriptionBean()
            }

            override fun newArray(size: Int): Array<SubscriptionBean?> {
                return arrayOfNulls(size)
            }
        }
    }

    var type: Int = 0
    var link: String = ""
    var token: String = ""
    var forceResolve: Boolean = false
    var deduplication: Boolean = false
    var filterNotRegex: String = ""
    var updateWhenConnectedOnly: Boolean = false
    var customUserAgent: String = ""
    var autoUpdate: Boolean = false
    var autoUpdateDelay: Int = 1440
    var lastUpdated: Int = 0

    // SIP008
    var bytesUsed: Long = 0L
    var bytesRemaining: Long = 0L // Also for OOC

    // Open Online Config
    var username: String = ""
    var expiryDate: Long = 0L

    override fun serializeToBuffer(output: ByteBufferOutput) {
        output.writeInt(4)

        output.writeInt(type)
        output.writeString(link)

        output.writeBoolean(forceResolve)
        output.writeBoolean(deduplication)
        output.writeBoolean(updateWhenConnectedOnly)
        output.writeString(customUserAgent)
        output.writeBoolean(autoUpdate)
        output.writeInt(autoUpdateDelay)
        output.writeInt(lastUpdated)
        output.writeLong(expiryDate)
        output.writeLong(bytesUsed)
        output.writeLong(bytesRemaining)
        output.writeString(token)
        output.writeString(filterNotRegex)
    }

    fun serializeForShare(output: ByteBufferOutput) {
        output.writeInt(1)

        output.writeInt(type)
        output.writeString(link)

        output.writeBoolean(forceResolve)
        output.writeBoolean(deduplication)
        output.writeBoolean(updateWhenConnectedOnly)
        output.writeString(customUserAgent)
        output.writeString(token)
    }

    override fun deserializeFromBuffer(input: ByteBufferInput) {
        val version = input.readInt()

        type = input.readInt()
        link = input.readString()
        forceResolve = input.readBoolean()
        deduplication = input.readBoolean()
        updateWhenConnectedOnly = input.readBoolean()
        customUserAgent = input.readString()
        autoUpdate = input.readBoolean()
        autoUpdateDelay = input.readInt()
        lastUpdated = input.readInt()

        if (version >= 2) {
            expiryDate = input.readLong()
            bytesUsed = input.readLong()
            bytesRemaining = input.readLong()
        }

        if (version >= 3) {
            token = input.readString()
        }

        if (version >= 4) {
            filterNotRegex = input.readString()
        }
    }

    fun deserializeFromShare(input: ByteBufferInput) {
        val version = input.readInt()

        type = input.readInt()
        link = input.readString()
        forceResolve = input.readBoolean()
        deduplication = input.readBoolean()
        updateWhenConnectedOnly = input.readBoolean()
        customUserAgent = input.readString()

        if (version >= 1) {
            token = input.readString()
        }
    }

}
